src/rendezvous/MSC3906Rendezvous.ts

```diff
@@ -104,39 +146,125 @@ export class MSC3906Rendezvous {
         // determine available protocols
         if (features.get(Feature.LoginTokenRequest) === ServerSupport.Unsupported) {
             logger.info("Server doesn't support MSC3882");
-            await this.send({ type: PayloadType.Finish, outcome: Outcome.Unsupported });
+            await this.send(
+                this.v1FallbackEnabled
+                    ? { type: PayloadType.Finish, outcome: Outcome.Unsupported }
+                    : { type: PayloadType.Failure, reason: FailureReason.Unsupported },
+            );
             await this.cancel(RendezvousFailureReason.HomeserverLacksSupport);
             return undefined;
         }
 
-        await this.send({ type: PayloadType.Progress, protocols: [LOGIN_TOKEN_PROTOCOL.name] });
+        await this.send({
+            type: this.v1FallbackEnabled ? PayloadType.Progress : PayloadType.Protocols,
+            protocols: [LOGIN_TOKEN_PROTOCOL.name],
+        });
 
         logger.info("Waiting for other device to chose protocol");
-        const { type, protocol, outcome } = await this.receive();
+        const nextPayload = await this.receive();
+
+        this.checkForV1Fallback(nextPayload);
+
+        const protocol = this.v1FallbackEnabled
+            ? await this.handleV1ProtocolPayload(nextPayload)
+            : await this.handleV2ProtocolPayload(nextPayload);
+
+        // invalid protocol
+        if (!protocol || !LOGIN_TOKEN_PROTOCOL.matches(protocol)) {
+            await this.cancel(RendezvousFailureReason.UnsupportedAlgorithm);
+            return undefined;
+        }
+
+        return checksum;
+    }
+
+    private checkForV1Fallback({ type }: MSC3906RendezvousPayload): void {
+        // even if we didn't start in v1 fallback we might detect that the other device is v1
+        if (type === PayloadType.Finish || type === PayloadType.Progress) {
+            // this is a PDU from a v1 flow so use fallback mode
+            this.v1FallbackEnabled = true;
+        }
+    }
```

Comment:
I'd be inclined to inline this method, but ymmv

Reply:
I had it inline but Sonarcloud complained that the function was too complicated, so I split it out!